source/Attribute.cpp

```diff
@@ -0,0 +1,425 @@
+{
+	if(effect == -1)
+		return false;
+	if(secondary != -1)
+	{
+		if(category != RESISTANCE)
+			return false;
+		if(effect >= ATTRIBUTE_EFFECT_COUNT || effect == PIERCING)
+			return false;
+		if(!Attribute(category, effect).IsSupported())
+			return false;
+		return secondary == ENERGY || secondary == HEAT || secondary == FUEL;
+	}
+	if(IsRelative())
+	{
+		if(IsMultiplier())
+			return false;
+		int effectType = effect % ATTRIBUTE_EFFECT_COUNT;
+		switch(category)
+		{
+			case DAMAGE:
+				if(effectType == DISABLED || effectType == MINABLE)
+					return true;
+			case FIRING:
+				if(effectType <= HULL || (effectType >= ENERGY && effectType <= HEAT))
+					return true;
+			default:
+				return false;
+		}
+	}
+	if(IsMultiplier())
+	{
+		if(category != SHIELD_GENERATION && category != HULL_REPAIR)
+			return false;
+		int effectType = effect % ATTRIBUTE_EFFECT_COUNT;
+		return effectType == ENERGY || effectType == HEAT || effectType == FUEL || effectType == category;
+	}
+	if(effect == THRUST && category == AFTERBURNING)
+		return true;
+	if(effect >= THRUST && effect <= CLOAK)
+		return static_cast<int>(effect) == static_cast<int>(category);
+	if(effect >= ENERGY && effect <= HEAT)
+		return (category != COOL && category != ACTIVE_COOL && category != RESISTANCE) ||
+				(effect == ENERGY && category == ACTIVE_COOL);
+	if(effect >= DISCHARGE && effect <= DISRUPTION)
+		return (category >= AFTERBURNING && category <= DAMAGE) || (category >= THRUSTING && category <= TURNING);
+	switch(effect)
+	{
+		case SHIELDS:
+		case HULL:
+			if(category <= HULL_REPAIR && static_cast<int>(effect) != static_cast<int>(category))
+				return false;
+			return (category < COOL || category > CLOAKING) && category != RESISTANCE;
+		case FORCE:
+			return category == FIRING || category == PROTECTION || category == DAMAGE;
+		case DISABLED:
+		case MINABLE:
+			return category == DAMAGE;
+		case PIERCING:
+			return category == DAMAGE || category == PROTECTION || category == RESISTANCE;
+		case DELAY:
+		case DEPLETED_DELAY:
+			return category == HULL_REPAIR || category == SHIELD_GENERATION;
+		case RAMSCOOP:
+			return category == PASSIVE;
+		default:
+			return false;
+	}
+}
+
+
+
+// Checks whether this attribute is a requirement for its category.
+// Required attributes mark resource consumption when an action is taken.
+bool Attribute::IsRequirement() const
+{
+	if(category == -1 || category == PASSIVE || category == DAMAGE || category == PROTECTION
+			|| category == COOL || effect == -1)
+		return false;
+	if(static_cast<int>(category) == static_cast<int>(effect) && category <= CLOAKING)
+		return false;
+	if(effect <= HULL || effect == ENERGY || effect == FUEL || effect == DELAY)
+		return true;
+	return false;
+}
+
+
+
+// Checks whether this attribute is a multiplier.
+bool Attribute::IsMultiplier() const
+{
+	return effect >= ATTRIBUTE_EFFECT_COUNT && (effect < ATTRIBUTE_EFFECT_COUNT * 2
+			|| effect >= ATTRIBUTE_EFFECT_COUNT * 3);
```

Comment:
```suggestion
		|| effect >= ATTRIBUTE_EFFECT_COUNT * 3);
```
one tab is enough

Reply:
Really? I thought we indented with 2 when continuing a line.

Reply:
you only do that if it's an if that has code on the next line